base-R.qmd

```diff
@@ -0,0 +1,502 @@
+
+knitr::include_graphics("images/pepper-1.jpg")
+```
+
+```{r}
+#| label: fig-pepper-3
+#| echo: false
+#| out-width: "25%"
+#| fig-cap: >
+#|   `pepper[[1]]`
+#| fig-alt: A photo of single packet of pepper.
+
+knitr::include_graphics("images/pepper-2.jpg")
+```
+
+This same principle applies when you use 1d `[` with a data frame:
+
+```{r}
+df <- tibble(x = 1:3, y = 3:5)
+
+# returns a one-column data frame
+df["x"]
+
+# returns the contents of x
+df[["x"]]
+```
+
+### Exercises
+
+1.  What happens when you subset with a positive integer that's bigger than the length of the vector?
+    What happens when you subset with a name that doesn't exist?
+
+2.  What would `pepper[[1]][1]` be?
+    What about `pepper[[1]][[1]]`?
+
+## Iteration
+
+In @sec-iteration, you learned tidyverse techniques for iteration, primarily `dplyr::across()` and `purrr::map()`.
+In this section, you'll learn about their base equivalents, first the apply family of functions and then the `for` loop, the powerful idea that underlies all these functions.
+
+Some people will tell you to avoid `for` loops because they are slow.
+They're wrong!
+(Well at least they're rather out of date, as `for` loops haven't been slow for many years.) The chief benefit of using functions like `map()` or `apply()` is not speed, but clarity: once you've mastered the basic idea, they make your code faster to write and easier to read.
+
+### Apply family
+
+Another way of saying "map a function over each element of a vector" is "apply a function over each element of a vector", and base R provides a family of functions known as the "apply" family because they all use apply as a suffix.
+Here we'll give you a quick overview of this family so you can recognize them in the wild.
+
+The most important member of this family is `lapply()`, which is very similar to `purrr::map()`[^base-r-3].
+In fact, because we don't use any of `map()`'s more advanced features, you can replace every `map()` call in @sec-iteration with `lapply()`.
+
+[^base-r-3]: It just lacks convenient features like progress bars and reporting which element caused the problem if there's an error.
+
+There's no exact base R equivalent to `across()` but you can get close by using `[` twice along with `lapply():`
+
+```{r}
+df <- tibble(a = 1, b = 2, c = "a", d = "b", e = 4)
+
+# First find numeric columns
+num_cols <- sapply(df, is.numeric)
+num_cols
+
+# Then transform each column with lapply() and replace the original values
+df[num_cols] <- lapply(df[num_cols], \(x) x * 2)
+df
+```
+
+The code above uses a new function, `sapply()`.
+It's similar to `lapply()` but it always tries to simplify the result, hence the `s` in its name, here producing a logical vector instead of a list.
+We don't recommend using it for programming, because the simplification can fail and give you an unexpected type, but it's usually fine for interactive use.
+
+Base R provides a stricter version of `sapply()` called `vapply()`, short for vector apply.
```

Comment:
Could draw some connections between `sapply->map_vec` and `vapply->map_int/dbl` if you wanted to